lib/inspec/input_registry.rb

```diff
@@ -166,15 +166,47 @@ def bind_inputs_from_cli_args(profile_name, input_list)
           end
         end
         input_name, input_value = pair.split("=")
+        input_value = parse_cli_input_value(input_name, input_value)
         evt = Inspec::Input::Event.new(
-          value: input_value.chomp(","), # Trim trailing comma if any
+          value: input_value,
           provider: :cli,
           priority: 50
         )
         find_or_register_input(input_name, profile_name, event: evt)
       end
     end
 
+    # Remove trailing commas, resolve type.
+    def parse_cli_input_value(input_name, given_value)
+      value = given_value.chomp(",") # Trim trailing comma if any
+      case value
+      when /^true|false$/
+        value = value == "true"
+      when /^-?\d+$/
+        value = value.to_i
+      when /^-?\d+\.\d+$/
+        value = value.to_f
+      when /^(\[|\{).*(\]|\})$/
```

Comment:
Possibly this should match on at least one syntax-ish character e.g. `'` `"` `,` or `:`
That would allow something that's a simple string with brackets to continue as a literal string e.g. `[DEBUG]` or `{EXPECTED_CI_VARIABLE}`

Related to this, the `raise` on line 203 stops users from intending a string that matches whatever we use for the  the pattern here.
It should possibly continue as a literal string to avoid raising an error on something the user didn't intend to be parsed. Flip side to that would be users intending JSON/YAML getting a string if (or _when_ for YAML :D ) they make a syntax error.

Reply:
But `--input=[a,b,c]` is valid a YAML array; it's one of the test cases. It contains none of `",',:`.

I can change the raise to a warn, if that's blocking acceptance.

Reply:
A warning feels better than an error, when the possibility is some edge case(s) with what the user's doing, thanks!